contrib/platform/src/com/sun/jna/platform/mac/IOKit.java

```diff
@@ -0,0 +1,546 @@
+    double kIOPSTimeRemainingUnknown = -1.0;
+
+    MachPort MACH_PORT_NULL = new MachPort();
+
+    /**
+     * IOKitLib implements non-kernel task access to common IOKit object types -
+     * IORegistryEntry, IOService, IOIterator etc. These functions are generic -
+     * families may provide API that is more specific.
+     * <p>
+     * IOKitLib represents IOKit objects outside the kernel with the types
+     * io_object_t, io_registry_entry_t, io_service_t, and io_connect_t. Function
+     * names usually begin with the type of object they are compatible with - e.g.,
+     * IOObjectRelease can be used with any io_object_t. Inside the kernel, the c++
+     * class hierarchy allows the subclasses of each object type to receive the same
+     * requests from user level clients, for example in the kernel, IOService is a
+     * subclass of IORegistryEntry, which means any of the IORegistryEntryXXX
+     * functions in IOKitLib may be used with io_service_t's as well as
+     * io_registry_t's. There are functions available to introspect the class of the
+     * kernel object which any io_object_t et al. represents. IOKit objects returned
+     * by all functions should be released with IOObjectRelease.
+     */
+    class IOObject extends PointerType {
+        public IOObject() {
+            super();
+        }
+
+        public IOObject(Pointer p) {
+            super(p);
+        }
+
+        /**
+         * Convenience method for {@link IOKit#IOObjectRelease} on this object.
+         *
+         * @return 0 if successful, otherwise a {@code kern_return_t} error code.
+         */
+        public int release() {
+            return INSTANCE.IOObjectRelease(this);
+        }
+    }
+
+    /**
+     * Communication between tasks is an important element of the Mach philosophy.
+     * Mach supports a client/server system structure in which tasks (clients)
+     * access services by making requests of other tasks (servers) via messages sent
+     * over a communication channel.
+     * <p>
+     * The endpoints of these communication channels in Mach are called ports, while
+     * port rights denote permission to use the channel.
+     */
+    class MachPort extends IOObject {
+        public MachPort() {
+            super();
+        }
+
+        public MachPort(Pointer p) {
+            super(p);
+        }
+    }
+
+    /**
+     * An IOKit iterator handle.
+     */
+    class IOIterator extends IOObject {
+        public IOIterator() {
+            super();
+        }
+
+        public IOIterator(Pointer p) {
+            super(p);
+        }
+
+        /**
+         * Convenience method for {@link IOKit#IOIteratorNext} on this object.
+         *
+         * @return If the iterator handle is valid, the next element in the iteration is
+         *         returned, otherwise zero is returned. The element should be released
+         *         by the caller when it is finished.
+         */
+        public IOService next() {
```

Comment:
> Can IOIterator only return IOServices? 

For all practical purposes, yes.  However, after looking at this in detail, I think it would be better to use `IORegistryEntry` as the return type.  I originally chose `IOService` because I obtained the iterator from calls to `IOServiceGetMatchingServices` or to the `IORegistryEntryGetChildEntry` and `IORegistryEntryGetParentEntry` calls using the `"IOService"` plane.

From a JNA library perspective, `IOKit` is mapping the mac `IOKit.framework` which borrows a few headers from the Kernel (`System.B.framework`).

From an IOKit perspective, the `IORegistryEntry` type is the base class, and probably a better choice here.  `IOService` handles nearly everything.  Apple docs say:
> All driver objects based on the I/O Kit inherit from the two base classes IORegistryEntry and IOService. The second of these classes, IOService, directly inherits from IORegistryEntry and all driver objects ultimately inherit from IOService. The IORegistryEntry class defines a driver object as a node in the I/O Registry, and IOService defines the life cycle of a driver object as well as implementing other behavior common to drivers.
>
> The close inheritance relationship between IORegistryEntry and IOService might invite speculation as to why these classes weren’t designed as one class. The reason is performance. Having IORegistryEntry as a superclass of IOService is an optimization because, in terms of memory footprint, the IORegistryEntry object is much more lightweight.

I almost got rid of the `IOService` class in my mapping, since I can use `IORegistryEntry` for all the registry lookup needs (and all uses of the iterator.)  But it's actually valuable when dealing with SMC calls where I need to open and close connections to the ports.

I can't find any examples where `io_iterator_t` is used for anything other than an `IORegistryEntry` (or subclass) iteration.  If they did, they would be in another framework that uses the header files and outside of the `IOKit` framework, so I think that for the `IOKit.IOIterator` class definition, in particular, it's safe to use `IORegistryEntry`.

Further support for use of `IORegistryEntry`:
> IOTypes.h seems to define all these objects in the same way, specifically, as mach_port_t objects. ... To understand why this is so, recall that a Mach port is the communication transport mechanism an application uses to communicate with the I/O Kit. As far as the kernel is concerned, every time your application communicates with any in-kernel object using one of the object types defined in IOTypes.h, it’s using the same generic mechanism (namely, a Mach port) to cross the user-kernel boundary.
>
> From the application’s point of view, however, the transport mechanism is unimportant and what matters is the type of object on the other side of the port. The fact that an io_iterator_t object, for example, encapsulates the association of a Mach port with an in-kernel IOIterator object is not as important to the application as *the fact that an io_iterator_t object refers to an in-kernel object that knows how to iterate over the I/O Registry.*

Reply:
Ok, sounds sane.